### PR TITLE
DrivAerML: 4L/384d seed sweep — 5 seeds to characterize initialization sensitivity

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The 4L/384d model that achieved 5.73% (PR #2602) was trained with a single default seed. Prior experiments on the 4L/256d architecture showed extreme seed sensitivity (seed=1 → 43.93% vs default → 12.70%). We don't yet know whether the wider 384d model shares this brittleness. Running 5 seeds characterizes the variance and tells us: (a) how reliable the 5.73% result is, (b) whether a better seed could push further without any architecture change, and (c) the true floor vs ceiling for this configuration.

## Instructions

Run 5 training jobs sequentially, each with a different random seed. Use the exact 4L/384d config from PR #2602 — change only `--seed` and `--wandb-name` for each run.

**Seed 1 (seed=42):**
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --seed 42 \
  --wandb-name "askeladd/drivaerml-4L384d-seed42" \
  --wandb_group "drivaerml-4L384d-seeds" \
  --agent askeladd
```

**Seed 2 (seed=123):**
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --seed 123 \
  --wandb-name "askeladd/drivaerml-4L384d-seed123" \
  --wandb_group "drivaerml-4L384d-seeds" \
  --agent askeladd
```

**Seed 3 (seed=456):**
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --seed 456 \
  --wandb-name "askeladd/drivaerml-4L384d-seed456" \
  --wandb_group "drivaerml-4L384d-seeds" \
  --agent askeladd
```

**Seed 4 (seed=789):**
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --seed 789 \
  --wandb-name "askeladd/drivaerml-4L384d-seed789" \
  --wandb_group "drivaerml-4L384d-seeds" \
  --agent askeladd
```

**Seed 5 (seed=1337):**
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --seed 1337 \
  --wandb-name "askeladd/drivaerml-4L384d-seed1337" \
  --wandb_group "drivaerml-4L384d-seeds" \
  --agent askeladd
```

Report all 5 `val_primary/surface_rel_l2_pct` results in your PR comment, plus mean ± std. Note the best seed — we may use it as the new default going forward.

## Baseline

- **val_primary/surface_rel_l2_pct**: 5.73% (PR #2602, epoch 144, W&B run `7ogfs7ph`)
- **Architecture**: 4L / 384d / 6H, Fourier features, no-EMA, AdamW lr=5e-4, cosine T_max=30
- **Budget**: 180-minute timeout, model was still improving at cutoff
- **Reproduce**: `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 --model-heads 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Seed Sweep Summary (4L/384d, 5 seeds)

| Seed | Best val_primary/surface_rel_l2_pct | Best Epoch | Final Epoch | test_surface/surface_rel_l2_pct | W&B Run ID |
|------|--------------------------------------|------------|-------------|----------------------------------|------------|
| 42   | **13.90%**                           | 35         | 35          | 14.00%                           | `r89qxvd9` |
| 123  | 14.25%                               | 35         | 35          | 13.97%                           | `n29fiutc` |
| 456  | 13.94%                               | 35         | 36          | 15.32%                           | `n5a0qo33` |
| 789  | 14.70%                               | 35         | 36          | 16.22%                           | `8c9m3los` |
| 1337 | 14.39%                               | 35         | 36          | 15.94%                           | `829c2i56` |

**Val best: 14.24% ± 0.33% (mean ± std)**
**Val range: 13.90% – 14.70% (spread = 0.80pp)**
**Test: 15.09% ± 1.06% (mean ± std)**
**Best seed by val: seed=42 (13.90%)**

### Comparison to Baseline

- **Baseline**: 5.73% (PR #2602, epoch 144, 180-minute timeout, run `7ogfs7ph`)
- **This sweep**: 14.24% ± 0.33% (best: 13.90%, 35-36 epochs, 30-minute timeout)

**Important caveat**: These runs were constrained to `SENPAI_MAX_EPOCHS=50` and `SENPAI_TIMEOUT_MINUTES=30`, reaching only 35-36 epochs before timeout. The original baseline ran for 144 epochs with a 180-minute budget. The model was still actively improving at cutoff in all 5 runs. These results characterize seed sensitivity at ~35 epochs of training, NOT at convergence.

### What Happened

1. **Low seed sensitivity at 35 epochs**: The 4L/384d architecture shows remarkably consistent behavior across seeds, with only 0.80pp spread (13.90%–14.70%) and std of 0.33%. This is a dramatic improvement over the 4L/256d architecture which showed >30pp spread between seeds (seed=1 → 43.93% vs default → 12.70%).

2. **All seeds on same trajectory**: Every seed hit its best val at epoch 35, and all training curves followed very similar patterns — rapid initial improvement, then steady convergence. The wider 384d model appears to stabilize training significantly.

3. **Val-test gap varies by seed**: Seed 123 had the tightest val-test gap (14.25% → 13.97%), while seed 789 had the largest (14.70% → 16.22%). This suggests some seeds generalize slightly better despite similar val performance.

4. **Best seed is 42**: Both the best val (13.90%) and competitive test (14.00%) come from seed=42. Seed=123 had the best test score (13.97%) with slightly worse val.

### Peak Memory
~5.7 GB GPU memory per run (observed via nvidia-smi during training).

### Commands Used
All 5 seeds ran in parallel on GPUs 0-4 with identical config, only varying `--seed`:
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=<N> python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 \
  --model-heads 6 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --seed <SEED> --wandb-name "askeladd/drivaerml-4L384d-seed<SEED>" \
  --wandb-group "drivaerml-4L384d-seeds" --agent askeladd
```

### Suggested Follow-ups

1. **Run seeds with longer budget**: To properly compare against the 5.73% baseline, these seeds need the same 180-minute / 999-epoch budget. The consistent training curves suggest all seeds would converge to similar final values.
2. **Adopt seed=42 as default**: It produced the best val result and has good val-test agreement. Could be used as the standard seed for future DrivAerML experiments.
3. **Test-time seed sensitivity**: The val-test gap variability (0.28pp for seed=123 vs 1.52pp for seed=789) warrants investigation — is this data sampling noise or genuine generalization differences?

### Note on PR Instructions
The PR body used `--wandb_group` (underscore) but train.py expects `--wandb-group` (hyphen). Fixed in the actual runs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2696 because ADVISOR: Closing — valuable seed sensitivity data but at 30-min budget (35-36 epochs), these results can't be compared to the 5.73% baseline (180-min, 151 epochs).

**KEY INSIGHTS RECORDED:**
1. 4L/384d has remarkably LOW seed sensitivity: 13.90-14.70% range (std=0.33%) vs 4L/256d's 30pp+ spread. Width scaling stabilizes training.
2. All 5 seeds converge to very similar trajectories — no lucky seeds at this width.
3. seed=42 was marginally best (13.90% val, 14.00% test).

These insights tell us: (a) DrivAerML at 4L/384d doesn't need multi-seed sweeps — one run is representative, and (b) further improvement must come from hyperparameters or architecture, not seed hunting.

askeladd, please pick up your next assignment..
